### PR TITLE
NUI editor autosave

### PR DESCRIPTION
**Pull request by rzats**
### Contains

Implements autosaving for the NUI editor, as per https://github.com/MovingBlocks/Terasology/pull/2438#issuecomment-241214770.

### How to test

Open the NUI (skin/screen) editor, edit, Alt+F4 without saving changes, restart the game and open the same editor again.

### Outstanding before merging

No outstanding issues.

### CODE REVIEW

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

File: engine/src/main/java/org/terasology/rendering/nui/editor/layers/AbstractEditorScreen.java [hunk 1]
```diff
@@ -113,6 +118,15 @@
     protected abstract Path getAutosaveFile();
 
     @Override
+    public void onDraw(Canvas canvas) {
+        super.onDraw(canvas);
+        if (!autosaveLoaded) {
+            loadAutosave();
+            autosaveLoaded = true;
+        }
```
Comment (flo) [inline engine/src/main/java/org/terasology/rendering/nui/editor/layers/AbstractEditorScreen.java]: That looks like a hack to me (no one would expect a autosave loading in a draw method). A comment why it is done there should at least be added.

**Comment by flo** [comment]
The autosave gets created and loaded, but it is neitehr possible to use the "Save" nor the "Save As" button. "Save" is grayed out and "Save As" results in an exception. This problem is propably related to the edited asset no longer being selected after you restarted the game

There is btw. also the special case that someone opens the editor in a situation where the loaded asset is no longer available. I leave it up to you if you want to handle it properly. It is not that important.

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by flo** [comment]
@rzats The "Save" button still isn't enabled (but "Save As") when you restart the game after a change.

I also noticed a second bug: When you select a different screen, but then choose to cancel the operation (since there are changes) then the save button will override the wrong widget. It will override the widget of which the selection got aborted and not the widget with the modification.

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by flo** [comment]
nice roundup, merged :)

**Pull request merged**